flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/gateway/context/SessionContextTest.java

```diff
@@ -0,0 +1,160 @@
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.apache.flink.configuration.CoreOptions.DEFAULT_PARALLELISM;
+import static org.apache.flink.configuration.PipelineOptions.NAME;
+import static org.apache.flink.table.api.config.TableConfigOptions.TABLE_SQL_DIALECT;
+
+/** Test {@link SessionContext}. */
+public class SessionContextTest {
+
+    private static final String DEFAULTS_ENVIRONMENT_FILE = "test-sql-client-defaults.yaml";
+
+    @Test
+    public void testSetAndResetYamlKey() throws Exception {
+        SessionContext sessionContext = createSessionContext();
+        sessionContext.set("execution.max-table-result-rows", "100000");
+        Assert.assertEquals(
+                "100000",
+                sessionContext
+                        .getSessionEnvironment()
+                        .getExecution()
+                        .asMap()
+                        .get("max-table-result-rows"));
+
+        sessionContext.reset();
+
+        Assert.assertEquals(
+                "100",
+                sessionContext
+                        .getSessionEnvironment()
+                        .getExecution()
+                        .asMap()
+                        .get("max-table-result-rows"));
+    }
+
+    @Test
+    public void testSetAndResetOption() throws Exception {
+        SessionContext sessionContext = createSessionContext();
+        // table config option
+        sessionContext.set(TABLE_SQL_DIALECT.key(), "hive");
+        // runtime config option and has default value
+        sessionContext.set(DEFAULT_PARALLELISM.key(), "128");
+        // runtime config option and doesn't have default value
+        sessionContext.set(NAME.key(), "test");
+        Assert.assertEquals(
+                "hive",
+                sessionContext
+                        .getSessionEnvironment()
+                        .getConfiguration()
+                        .asMap()
+                        .get(TABLE_SQL_DIALECT.key()));
+        Assert.assertEquals(
+                "hive",
+                sessionContext
+                        .getExecutionContext()
+                        .getTableEnvironment()
+                        .getConfig()
+                        .getConfiguration()
+                        .getString(TABLE_SQL_DIALECT));
+        Assert.assertEquals(
+                128,
+                sessionContext
+                        .getExecutionContext()
+                        .getTableEnvironment()
+                        .getConfig()
+                        .getConfiguration()
+                        .getInteger(DEFAULT_PARALLELISM));
+        Assert.assertEquals(
+                "test",
+                sessionContext
+                        .getExecutionContext()
+                        .getTableEnvironment()
+                        .getConfig()
+                        .getConfiguration()
+                        .getString(NAME));
+
+        sessionContext.reset();
```

Comment:
Could you try to reset configuration in `flink-conf.yaml` (i.e. the `Configuration` in `DefaultContext`)? According to the desgin, the configuration can't be reset.